store/lookup/table.go

```diff
@@ -0,0 +1,161 @@
+// In the case of a collision, more bytes of the hash will be used for disambiguation but this happens
+// in a minority of cases except for massively large data sets.
+func NewTable(prefix []byte) (Table, error) {
+	return NewTableWithOptions(TableOptions{
+		Prefix: prefix,
+	})
+}
+
+// NewTableWithOptions creates a Table with custom options. Most users should just use NewTable
+// with the default values.
+func NewTableWithOptions(options TableOptions) (Table, error) {
+	prefixLen := len(options.Prefix)
+	minLength := options.MinLength
+	if minLength == 0 {
+		minLength = 4
+	}
+
+	newHash := options.NewHash
+	if newHash == nil {
+		newHash = func() hash.Hash {
+			return fnv.New64a()
+		}
+	}
+
+	hashLen := len(newHash().Sum(nil))
+	if minLength > hashLen {
+		return nil, fmt.Errorf("option MinLength %d is greater than hash length %d", minLength, hashLen)
+	}
+
+	bufLen := prefixLen + hashLen + binary.MaxVarintLen64
+	initLen := prefixLen + minLength
+
+	return table{
+		minLen:    minLength,
+		bufLen:    bufLen,
+		initLen:   initLen,
+		newHash:   newHash,
+		prefix:    options.Prefix,
+		prefixLen: prefixLen,
+		hashLen:   hashLen,
+	}, nil
+}
+
+// TableOptions is used to specify custom table options and should only be used by advanced users.
+type TableOptions struct {
+	// NewHash is a function which returns a new hash.Hash instance.
+	NewHash func() hash.Hash
+
+	// MinLength is the minimum number of hash bytes that will be used to create a lookup identifier.
+	MinLength int
+
+	// Prefix is an optional prefix to be pre-pended to all KVStore keys.
+	Prefix []byte
+}
+
+type table struct {
+	minLen    int
+	bufLen    int
+	newHash   func() hash.Hash
+	prefix    []byte
+	prefixLen int
+	hashLen   int
+	initLen   int
+}
+
+// KVSTore is the interface for key-value stores that Tables operate on.
+type KVStore interface {
+	// Get returns nil iff key doesn't exist. Panics on nil key.
+	Get(key []byte) []byte
+
+	// Set sets the key. Panics on nil key or value.
+	Set(key, value []byte)
+}
+
+func (t table) GetValue(store KVStore, id []byte) []byte {
+	buf := make([]byte, t.prefixLen+len(id))
+	copy(buf, t.prefix)
+	copy(buf[t.prefixLen:], id)
+	return store.Get(id)
+}
+
+func (t table) GetOrCreateID(store KVStore, value []byte) []byte {
+	id, _ := t.getOrCreateID(store, value)
+	return id
+}
+
+func (t table) getOrCreateID(store KVStore, value []byte) (id []byte, numCollisions int) {
+	hasher := t.newHash()
+	_, err := hasher.Write(value)
+	if err != nil {
+		// we panic here because hash.Write returning an error shouldn't happen
+		panic(err)
+	}
+	hashBz := hasher.Sum(nil)
+
+	id = make([]byte, t.initLen, t.bufLen)
+	copy(id, t.prefix)
+	copy(id[len(t.prefix):], hashBz[:t.minLen])
+	// take the first i bytes of hashBz starting with t.minLen and increasing
+	// in cases where there are collisions
+
+	for i := t.minLen; ; i++ {
+		if tryID(store, id, value) {
+			return id, i - t.minLen
+		}
+
+		if i >= t.hashLen {
+			break
+		}
```

Comment:
the 3 lines above can be moved to the for condition, because at the start of the loop we always have `i < t.minLen`
```
for i := t.minLen; i<t.hashLen; i++ {
```

Reply:
at the start of the loop `i == t.minLen` and if `t.minLen == t.hashLen` we have array out of bounds